### ISSUE TITLE
Cannot load additional resources for HTML page loaded from local on-demand resource

### ISSUE THREAD

**Issue by jfred1979** [react-native-community/react-native-webview#1179]
**Bug description:**

I'm attempting to load a local website which has been downloaded as an on-demand resource bundle. On the native side I load the bundle, then pass the path of the index file back to React Native, which is then used by a RN Webview to load it. The index file loads successfully, but additional resources loaded by that page are being blocked with this error being thrown in the Javascript console:
`Failed to load resource: The operation couldn’t be completed. Operation not permitted`

The current WebView tag looks like this:
`<WebView
            javaScriptEnabled={true}
            webkit={false}
            source={{ uri: indexURI }}
            style={{ flex: 1 }}
            allowFileAccess={true}
            originWhitelist={['*']} 
          />
`

The index file I am loading is extremely simple, just the required HTML and body tags and an image tag with './test.png' as the source

Am I missing a property in the WebView that would allow these assets to be loaded?

**Expected behavior:**
HTML files should be loaded along with all of the resources that it attempts to load

**Environment:**
 - OS: iOS
 - OS version: 12.4.4
 - react-native version: 0.59.9
 - react-native-webview version: 8.0.5

**Comment by flashman2** [comment]
Nothing is missed, its just a WKWebView issue mentioned here:
https://github.com/react-native-community/react-native-webview/issues/312#issuecomment-569327647

Really hope it will b fixed soon, if anyone have a solution (modifications in native code) please share! We get stuck with this too on iOS. However, on Android webview works fine.

**Comment by gdmerm** [comment]
@flashman2 Hi, there's a workaround you can try. Load your index.html file as you do now but also spawn an embedded static server to serve all the static assets being requested by the webview through that static server.

You can have a look at the "react-native-static-server" plugin that uses GCDWebServer on iOS to serve static assets. Leveraging the GCDWebServer is actually easy enough to even write your own native module around it, should the react-native-static-server does not fulfil all your needs. (in most cases it will though).

**Comment by flashman2** [comment]
@gdmerm Thanks, we will try this workaround.

**Comment by ekkis** [comment]
@gdmerm isn't that expensive from a CPU and app size perspectives?

**Comment by gdmerm** [comment]
@ekkis I always advise to profile any solution using either xcode or flipper. It also depends on the kind of devices your target audience use etc.

Should you find that you need to optimize (but again profile first, then optimize), then there are a few things you can do.

1. Stop the embedded server once the DOMContentLoaded event is fired (or catch the associated WebView event).

Once you have all your files loaded into WebView cache, you no longer need the server. There is an exception to this and that is if your HTML5 code sends ajax requests back to the domain that it originated from (that would you be your server). If you are issuing requests to other domains using CORS then you should still be ok.

2. If you find that your HTML5 codebase is issuing ajax requests back to your server, then you could possibly modify your codebase to instead issue postMessage() events and have the react-native side catch those events and issue the requests on behalf of your HTML5 code. That will still allow you to terminate the server once you receive the DOMContentLoaded event.

3. Should you find that you absolutely cannot stop the server for any reason in an early fashion as described above, then I suggest trying to suspend the server when your app is in the background and restart it when it's brought back to the foreground and then profile again. You can easily do that either by handling the associated react-native events or by enabling the `GCDWebServerOption_AutomaticallySuspendInBackground` option on GCDWebServer.

But in all terms n purposes I can't really tell you how expensive the overall solution will be. That is a factor of what your requirements are, what devices and OS versions you are targeting, how long your users spend on the WebView and how smart you can be working around any issues. A POC using the embedded server should be very quick to setup with barebone functionality. I suggest you do that and profile, profile, profile.

**Comment by ekkis** [comment]
awesome response.  thank you for taking the time to articulate it.  incidentally, you mention the WebView cache.  I have turned it off but am having problems where the page it displays is old.  I've verified that the content the server returns is correct but what gets rendered is a previous version.  turning off the cache doesn't seem to solve this problem, at least in iOS.  I'm getting set up with Android to see if I can replicate the issue there as well

**Comment by marconett** [comment]
Loading local content with `source={ uri: <uri> }` does work. It was fixed in #771.
You need to set `allowingReadAccessToURL={'file://'+<path_to_resource_folder>}`.

What's still a problem, is loading local content if you import a html string into a webview via `source={ html: <html_string>, baseUrl: <path_to_resource_folder> }`. In this case, setting `allowingReadAccessToURL` has no effect, as the underlying WKWebView doesn't support that paramater paired with loading HTML from a string.
This problem does not show up on a simulator, or if `<path_to_resource_folder>` points to the apps bundle folder.

This seems to be a deeper issue with WKWebView itself (https://bugs.webkit.org/show_bug.cgi?id=154916) and I am not sure how this library is supposed to fix this.

But working around the problem can be achieved in a couple of ways:
- Loading local content from as local webserver (as @gdmerm described)
- Writing the html string to a temp file (for example: `fs.writeFile(<path_to_temp_file>, <html>, 'utf8');`) and then using `source={ uri: <path_to_temp_file> }` and `allowingReadAccessToURL={<path_to_resource_folder>}` (tested this, it works).
- Rewriting all local content references to a custom url scheme and writing a iOS native custom url handler (https://bugs.webkit.org/show_bug.cgi?id=154916#c28)
- Parsing the HTML in JS and replacing all references to local content with inline alternatives

**Comment by ekkis** [comment]
@marconett I'm using the local web server approach but finding that images don't get loaded properly.  interestingly, other references (e.g. `<script src="..."></script>` and `<link rel="stylesheet" href="..."/>` work fine but when I have a `<img src="file.ext"/>` or `<img src="./file.ext"/>` I only get a little blue image with a question mark.  I'm not currently using `allowingReadAccessToURL` but I'm reasoning that if I don't need it for scripts and stylesheets, I shouldn't need it for images

any thoughts as to what might be wrong (happy to share code if needed)?

**Comment by ekkis** [comment]
@marconett incidentally, I abandoned the local webserver approach and tried the local file url.  I initially set the `allowingReadAccessToURL` as you recommend but found it didn't work, then found that `originWhitelist={["file://"]}` fixes the problem but I still can't see them images

**Comment by ekkis** [comment]
incidentally, it fails in the same way for both iOS and Android

**Comment by marconett** [comment]
please share the related lines of code, so i can try to help.

**Comment by mightyorange** [comment]
@marconett 
Thanks, Your solution just saved my life.

I even tried to place the `index.html` along with CSS file to App Bundle directory(https://medium.com/@caphun/react-native-load-local-static-site-inside-webview-2b93eb1c4225), 
then let `allowingReadAccessToURL ` point to App Document directory, not working also ---- `operation not permitted`

Finally i tried your solution ---- write all html file to App Document directory on the runtime,  let `allowingReadAccessToURL` keep point to App Document,  it works!!

**Comment by ekkis** [comment]
sorry for the long delay.  just getting get back to this... here's my codebase, for the expo-ejected project
```js
import React from 'react'
import {
  Text, StyleSheet,
  SafeAreaView, View, TextInput, Image,
  TouchableHighlight
} from 'react-native'
import {WebView} from 'react-native-webview'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import {
  faCog, faArrowAltCircleRight, faStar
} from '@fortawesome/free-solid-svg-icons'
import lib from './my-library'
import fs from 'react-native-fs'
import StaticServer from 'react-native-static-server'

const DEBUG = true

class App extends React.Component {
  constructor(props) {
    super(props)
    this.fetch = this.fetch.bind(this)
    this.dir = this.dir.bind(this)
    this.state = {html: '', baseUrl: '', www: ''}
  }
  async componentDidMount() {
    log('-- starting server')
    debug('Path=' + root())
    this.server = new StaticServer(8080, root(), {localOnly: true})
    var www = await this.server.start()
    debug('Listening on: ' + www)
    this.setState({www})
    // moveAndroidFiles()
    var ls = await fs.readDir('/data/user/0/com.browser/files/www')
    return console.log(ls.map(r => r.name))

  }
  async componentWillUnmount() {
    log('-- stopping server')
    if (this.server && await this.server.isRunning())
      this.server.stop()
  }
  async fetch() {
    var {url} = this.state
    await lib.site(url, root())
    this.setState({uri: this.state.www + '/index.html'})
    var html = await fs.readFile(root() + 'index.html', 'utf8')
    // var baseUrl = uri.replace(/\/[^/]+$/, '')
    this.setState({html})
  }
  async dir() {
    var ls = await fs.readDir(root())
    var ret = []
    for (var i = 0; i < ls.length; i++)
      ret.push(await fs.stat(ls[i].path))
    console.log(ret.map(r => ({mode: r.mode, path: r.path})))
    console.log(ls.map(r => r.name))
    this.setState({uri: this.state.www + '/index.html'})
    // var res = await fetch(this.state.www + '/index.html')
    // console.log(await res.text())
  }
  render() {
    return (
      <SafeAreaView style={styles.SafeAreaView}>
        <View style={styles.Toolbar}>
          <TouchableHighlight onPress={() => alert('cog')}>
            <FontAwesomeIcon icon={faCog} size={22} style={{marginRight: 5}} />
          </TouchableHighlight>
          <TouchableHighlight onPress={this.dir}>
            <FontAwesomeIcon icon={faStar} size={22} />
          </TouchableHighlight>
          <TextInput 
            style={styles.Url}
            placeholder="eos://<domain>" 
            onChangeText={val => this.state.url = val}
            autoCapitalize="none"
            />
          <TouchableHighlight onPress={this.fetch}>
            <FontAwesomeIcon icon={faArrowAltCircleRight} size={22} />
          </TouchableHighlight>
        </View>
        {/* <Text>
          {this.state.uri}
          {'\n' + this.state.html}
        </Text> */}
        {/* <Image
          style={{width: 250, height: 250, borderColor: 'red', borderWidth: 1}}
          source={{
            uri: Platform.OS === 'ios'
              ? (this.state.www + '/logo.jpg')
              : `file://${this.state.baseUrl + '/logo.jpg'}`,
          }}
        /> */}
        <WebView 
          // source={this.state.index ? require(this.state.uri) : {html: '<font size="60pt">Loading...</font>'}} 
          // source={{html: `<font size="60pt">Loading...${'index' in this.state ? this.state.index : 'not defined'}</font>`}} 
          // source={{html: this.state.status}}
          // source={{html: this.state.html, baseUrl: 'file://'}}
          source={{uri: this.state.uri}}
          style={styles.WebView}
          // originWhitelist={['*']}
          // allowUniversalAccessFromFileURLs={true}
          // allowFileAccessFromFileURLs={true}
          // allowFileAccess={true}
          cacheEnabled={false}
          useWebKit={true}
          />
      </SafeAreaView>
    )
  }
}

function root(dir = '/www/') {
  var path = Platform.OS === "android"
      ? fs.DocumentDirectoryPath
      : fs.MainBundlePath
  return path + dir
}

async function moveAndroidFiles() {
  if (Platform.OS === "android") {
    var ls = await fs.readDir(fs.DocumentDirectoryPath + '/www')
    console.log('MOVE', fs.DocumentDirectoryPath, ls.map(r => r.name))

    await fs.mkdir(fs.DocumentDirectoryPath + "/www");
    const files = ["www/index.html", "www/index.css", "www/index.js", "www/logo.jpg"];
    await files.forEach(async file => {
      await fs.copyFileAssets(file, fs.DocumentDirectoryPath + "/" + file);
    });
  }
}

function log() {
  console.log.apply(null, arguments)
}

function debug() {
  if (DEBUG) log.apply(null, arguments)
}

const styles = StyleSheet.create({
  SafeAreaView: {
    flex: 1,
    borderColor: 'red',
    borderWidth: 0
  },
  Toolbar: {
    paddingHorizontal: 12,
    paddingVertical: 12,
    borderWidth: 1,
    flexDirection: 'row'
  },
  Tool: {
    width: 20, height: 20, flex: 0
  },
  Url: {
    marginLeft: 10,
    marginRight: 6,
    borderWidth: 1,
    paddingHorizontal: 5,
    paddingVertical: 1,
    flex: 1,
    fontSize: 16
  },
  WebView: {
    borderColor: 'red', 
    borderStyle: "solid",
    borderWidth: 1, 
    flex: 1,
    alignContent: 'stretch',
    height: '100%'
  },
});

export default App;
```
please consider the above code as purely experimental.  the './my-library' above exports a function `site` that's responsible for writing 4 files to the directory provided.  these files look like this:
-- index.html --
```html
<html>
  <head>
    <link rel="stylesheet" type="text/css" href="index.css" />
  </head>
  <body>
    <p id="text">Hello World!</p>
	<img src="logo.jpg">
    <div id="newStuff"></div>
    <script type="text/javascript" src="./index.js"></script>
  </body>
</html>
```
-- index.css --
```css
#text {
  color: red;
}
```
-- index.js --
```js
document.getElementById("newStuff").innerHTML = "The javascript works!";
```
and a small image called `logo.jpg` (I've tried .png too with equal results)

**Comment by ekkis** [comment]
the code above displays the page and properly sources the CSS and Javascript such that I get a red "Hello World!" and notice that the javascript works, but the image doesn't display.  I get a little icon with a ? in blue instead.  I've also tried `src="./logo.jpg"` but it fails as well

**Comment by marconett** [comment]
you're still using the local webserver, i can't help you with that.

otherwise, you already seem to have your html as a file. all you should need to do is get rid of the local webserver and set `allowingReadAccessToURL={<path_to_folder_containing_html_file>}`.

**Comment by ekkis** [comment]
so in that case I would use `source={ uri: "file://..." }`?

**Comment by marconett** [comment]
Yes

**Comment by ekkis** [comment]
ok.  here's the cleaned up code:
```js
import React from 'react'
import {
  Text, StyleSheet,
  SafeAreaView, View, TextInput, Image,
  TouchableHighlight
} from 'react-native'
import {WebView} from 'react-native-webview'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import {
  faCog, faArrowAltCircleRight, faStar
} from '@fortawesome/free-solid-svg-icons'
import lib from './my-library'
import fs from 'react-native-fs'

class App extends React.Component {
  constructor(props) {
    super(props)
    this.fetch = this.fetch.bind(this)
    this.state = {uri: ''}
  }
  async fetch() {
    var {url} = this.state
    await lib.site(url, root())
    this.setState({uri: 'file://' + root() + 'index.html'})
    console.log(this.state.uri)
  }
  render() {
    return (
      <SafeAreaView style={styles.SafeAreaView}>
        <View style={styles.Toolbar}>
          <TextInput 
            style={styles.Url}
            placeholder="eos://<domain>" 
            onChangeText={val => this.state.url = val}
            autoCapitalize="none"
            />
          <TouchableHighlight onPress={this.fetch}>
            <FontAwesomeIcon icon={faArrowAltCircleRight} size={22} />
          </TouchableHighlight>
        </View>
        <WebView 
          source={{uri: this.state.uri}}
          style={styles.WebView}
          allowingReadAccessToURL={'file://' + root()}
          />
      </SafeAreaView>
    )
  }
}

function root(dir = '/www/') {
  var path = Platform.OS === "android"
      ? fs.DocumentDirectoryPath
      : fs.MainBundlePath
  return path + dir
}
```
when `lib.site()` executes, it prints the names of the files it writes to disk.  here's the output I get:

>  LOG  /Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/index.css
>  LOG  /Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/index.html
>  LOG  /Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/logo.png
>  LOG  /Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/index.js
>  LOG  file:///Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/index.html
>  WARN  Error opening URL:  [Error: Unable to open URL: file:///Users/ekkis/Library/Developer/CoreSimulator/Devices/9F96B01B-3C66-4ABA-96A5-B58049FB9212/data/Containers/Bundle/Application/DA124337-7414-45B0-ADDD-63163A9A6CB9/browser.app/www/index.html]

and the last entry is from the log in my fetch() function.  the page never opens on the simulator.  ideas?

**Comment by ekkis** [comment]
incidentally, if I add these lines to the end of my `fetch()` function:
```js
    var html = await fs.readFile(root() + 'index.html', 'utf8')
    console.log(html)
```
it prints the contents of the file so I am guaranteed the file exists

**Comment by ekkis** [comment]
hmm... adding this to the tag seems to work:
```js
originWhitelist={['*']}
```
however, it presents another problem (I had already gone down this path before I implemented the local webserver): when I request one url, it displays the file.  if I request a different, it continues to display the same file.  it doesn't matter what I request, I always get the same file.  I tried turning the cache off, but to no avail:
```js
cacheEnabled={false}
```
